finality-aleph/src/sync/data.rs

```diff
@@ -0,0 +1,97 @@
+use std::mem::size_of;
+
+use aleph_primitives::MAX_BLOCK_SIZE;
+use codec::{Decode, Encode, Error as CodecError, Input as CodecInput};
+use log::warn;
+
+use crate::{sync::Justification, Version};
+
+/// The representation of the database state to be sent to other nodes.
+/// In the first version this only contains the top justification.
+#[derive(Clone, Debug, Encode, Decode)]
+pub struct State<J: Justification> {
+    top_justification: J::Unverified,
+}
+
+/// Data to be sent over the network.
+#[derive(Clone, Debug, Encode, Decode)]
+pub struct NetworkData<J: Justification> {
+    state: State<J>,
+}
+
+/// Version wrapper around the network data.
+#[derive(Clone, Debug)]
+pub enum VersionedNetworkData<J: Justification> {
+    // Most likely from the future.
+    Other(Version, Vec<u8>),
+    V1(NetworkData<J>),
+}
+
+// We need 32 bits, since blocks can be quite sizeable.
+type ByteCount = u32;
+
+// We want to be able to safely send at least 10 blocks at once, so this gives uss a bit of wiggle
+// room.
+const MAX_SYNC_MESSAGE_SIZE: u32 = MAX_BLOCK_SIZE * 11;
+
+fn encode_with_version(version: Version, payload: &[u8]) -> Vec<u8> {
+    let size = payload.len().try_into().unwrap_or(ByteCount::MAX);
+
+    if size > MAX_SYNC_MESSAGE_SIZE {
+        warn!(
+            "Versioned sync message v{:?} too big during Encode. Size is {:?}. Should be {:?} at max.",
+            version,
+            payload.len(),
+            MAX_SYNC_MESSAGE_SIZE
+        );
+    }
+
+    let mut result = Vec::with_capacity(version.size_hint() + size.size_hint() + payload.len());
+
+    version.encode_to(&mut result);
+    size.encode_to(&mut result);
+    result.extend_from_slice(payload);
+
+    result
+}
+
+impl<J: Justification> Encode for VersionedNetworkData<J> {
+    fn size_hint(&self) -> usize {
+        use VersionedNetworkData::*;
+        let version_size = size_of::<Version>();
+        let byte_count_size = size_of::<ByteCount>();
+        version_size
+            + byte_count_size
+            + match self {
+                Other(_, payload) => payload.len(),
+                V1(data) => data.size_hint(),
+            }
+    }
+
+    fn encode(&self) -> Vec<u8> {
+        use VersionedNetworkData::*;
+        match self {
+            Other(version, payload) => encode_with_version(*version, payload),
+            V1(data) => encode_with_version(Version(1), &data.encode()),
+        }
+    }
+}
+
+impl<J: Justification> Decode for VersionedNetworkData<J> {
+    fn decode<I: CodecInput>(input: &mut I) -> Result<Self, CodecError> {
+        use VersionedNetworkData::*;
+        let version = Version::decode(input)?;
+        let num_bytes = ByteCount::decode(input)?;
+        match version {
+            Version(1) => Ok(V1(NetworkData::decode(input)?)),
+            _ => {
+                if num_bytes > MAX_SYNC_MESSAGE_SIZE {
```

Comment:
shouldnt this be checked before `match version`? Actually shouldn't we return immediately if the input is too large? Or is it even a good place for the such check?

Reply:
The substrate network allocates a lot of space for the message anyway, as it [returns the whole vector](https://github.com/Cardinal-Cryptography/aleph-node/blob/23935d233450cff25dd40fe7922f605f598469bc/finality-aleph/src/network/substrate.rs#L231), so doing this doesn't really help. And for messages that we can decode there is no much reason to drop them here, we will drop them soon if they are incorrect anyway. This is mostly so we don't allocate too much space based on a malicious wrong-version message in the buffer a couple lines below – a malicious actor could then send an extremely short message containing just a ridiculous version number followed by `u32::MAX` and make us allocate more memory than we have.